query-graphs-js/src/querygraph.ts

```diff
@@ -664,42 +664,54 @@ function federateSubgraphs(supergraph: Schema, subgraphs: QueryGraph[]): QueryGr
+              // Note that later, when we've handled @provides, this might not be true anymore as @provides may create copy of a
+              // certain type. But for now, it's true.
+              assert(
+                otherVertices.length == 1,
+                () => `Subgraph ${j} should have a single vertex for type ${type.name} but got ${otherVertices.length}: ${inspect(otherVertices)}`);
+
+              const otherVertice = otherVertices[0];
+              // The edge goes from the otherSubgraph to the current one.
+              const head = copyPointers[j].copiedVertex(otherVertice);
+              const tail = copyPointers[i].copiedVertex(v);
+              builder.addEdge(head, tail, new KeyResolution(), conditions);
             }
-            // Note that later, when we've handled @provides, this might not be true anymore as @provides may create copy of a
-            // certain type. But for now, it's true.
-            assert(
-              otherVertices.length == 1,
-              () => `Subgraph ${j} should have a single vertex for type ${type.name} but got ${otherVertices.length}: ${inspect(otherVertices)}`);
-
-            const otherVertice = otherVertices[0];
-            // The edge goes from the otherSubgraph to the current one.
-            const head = copyPointers[j].copiedVertex(otherVertice);
-            const tail = copyPointers[i].copiedVertex(v);
-            builder.addEdge(head, tail, new KeyResolution(), conditions);
-
-            // Additionally, if the key is on an @interfaceObject and this "other" subgraph has the type as
-            // a proper interface, then we need an edge from each of those implementation (to the @interfaceObject).
+
+            // Additionally, if the key is on an @interfaceObject and this "other" subgraph has some of the implementations
+            // of the corresponding interface, then we need an edge from each of those implementations (to the @interfaceObject).
             // This is used when an entity of specific implementation is queried first, but then some of the
             // requested fields are only provided by that @interfaceObject.
-            const otherType = otherVertice.type;
-            if (isInterfaceObject && isInterfaceType(otherType)) {
-              for (const implemType of otherType.possibleRuntimeTypes()) {
-                // Note that we're only taking the implementation types from "otherSubgraph", so we're guaranteed
-                // to have a corresponding vertice (and only one for the same reason than mentioned in the assert above).
-                const implemVertice = otherSubgraph.verticesForType(implemType.name)[0];
+            if (isInterfaceObject) {
+              const typeInSupergraph = supergraph.type(type.name);
+              assert(typeInSupergraph && isInterfaceType(typeInSupergraph), () => `Type ${type} is an interfaceObject in subgraph ${i}; should be an interface in the supergraph`);
```

Comment:
Assuming that these get wrapped in an Error further up?

Reply:
I'm not sure to understand what "these" refers to. But `assert` throws an `Error` if it ever does anything, so unsure why there would need to be additional wrapping further up?